ml4gw/waveforms/taylorf2.py

```diff
@@ -0,0 +1,165 @@
+    pfa_v1 *= pfaN
+    pfa_v2 = 5.0 * (74.3 / 8.4 + 11.0 * eta) / 9.0
+    pfa_v2 *= pfaN
+    pfa_v3 = -16.0 * PI
+    pfa_v3 *= pfaN
+    pfa_v4 = (
+        5.0
+        * (3058.673 / 7.056 + 5429.0 / 7.0 * eta + 617.0 * eta * eta)
+        / 72.0
+    )
+    pfa_v4 *= pfaN
+    pfa_v5logv = 5.0 / 3.0 * (772.9 / 8.4 - 13.0 * eta) * PI
+    pfa_v5logv *= pfaN
+    pfa_v5 = 5.0 / 9.0 * (772.9 / 8.4 - 13.0 * eta) * PI
+    pfa_v5 *= pfaN
+    pfa_v6logv = -684.8 / 2.1
+    pfa_v6 = (
+        11583.231236531 / 4.694215680
+        - 640.0 / 3.0 * PI * PI
+        - 684.8 / 2.1 * GAMMA
+        + eta * (-15737.765635 / 3.048192 + 225.5 / 1.2 * PI * PI)
+        + eta * eta * 76.055 / 1.728
+        - eta * eta * eta * 127.825 / 1.296
+        + pfa_v6logv * torch.log(torch.tensor(4.0))
+    )
+    pfa_v6logv *= pfaN
+    pfa_v6 *= pfaN
+    pfa_v7 = PI * (
+        770.96675 / 2.54016 + 378.515 / 1.512 * eta - 740.45 / 7.56 * eta * eta
+    )
+    pfa_v7 *= pfaN
+    # construct power series
+    phasing = (v7.T * pfa_v7).T
+    phasing += (v6.T * pfa_v6 + v6_logv.T * pfa_v6logv).T
+    phasing += (v5.T * pfa_v5 + v5_logv.T * pfa_v5logv).T
+    phasing += (v4.T * pfa_v4).T
+    phasing += (v3.T * pfa_v3).T
+    phasing += (v2.T * pfa_v2).T
+    phasing += (v1.T * pfa_v1).T
+    phasing += (v0.T * pfa_v0).T
+    # Divide by 0PN v-dependence
+    phasing /= v5
+
+    return phasing
+
+
+def taylorf2_amplitude(f: TensorType, mass1, mass2, distance) -> TensorType:
+    mass1_s = mass1 * MTSUN_SI
+    mass2_s = mass2 * MTSUN_SI
+    M_s = mass1_s + mass2_s
+    eta = mass1_s * mass2_s / M_s / M_s
+    Mf = (f.T * M_s).T
+    v = (PI * Mf) ** (1.0 / 3.0)
+    v10 = v**10
+
+    # Flux and energy coefficient at newtonian
+    FTaN = 32.0 * eta * eta / 5.0
+    dETaN = 2 * (-eta / 2.0)
+
+    amp0 = -4.0 * mass1_s * mass2_s * (PI / 12.0) ** 0.5
+
+    amp0 /= distance * MPC_SEC
+    flux = (v10.T * FTaN).T
+    dEnergy = (v.T * dETaN).T
+    amp = torch.sqrt(-dEnergy / flux) * v
+    amp = (amp.T * amp0).T
+
+    return amp
+
+
+def taylorf2_htilde(f: TensorType, params: TensorType, f_ref: float):
+    mass1 = params[:, 0]
+    mass2 = params[:, 1]
+    distance = params[:, 2]
+    phic = params[:, 3]
+
+    # repeat freq across batch size
+    f = f.repeat([mass1.shape[0], 1])
+    f_ref = torch.tensor(f_ref).repeat([mass1.shape[0], 1])
+
+    Psi = taylorf2_phase(f, mass1, mass2)
+    Psi_ref = taylorf2_phase(f_ref, mass1, mass2)
+
+    Psi = (Psi.T - 2 * phic).T
+    Psi -= Psi_ref
+
+    amp0 = taylorf2_amplitude(f, mass1, mass2, distance)
+    h0 = amp0 * torch.exp(-1j * (Psi - PI / 4))
+    return h0
+
+
+def TaylorF2(f: TensorType, params: TensorType, f_ref: float):
+    """
+    TaylorF2 up to 3.5 PN in phase. SPA amplitude.
+    params = [mass1, mass2, D, phic, inclination]
+
+    Returns:
+    --------
+      hp, hc
+    """
+    # shape assumed (n_batch, params)
+    # frequency array is repeated along batch
+    inclination = params[:, 4]
```

Comment:
I think it would be cleaner to have `params` be a dictionary with the parameter names as keys

Reply:
What can be done is to pass explicit variables mass1, mass2, etc. But each should have the shape (n_batch, 1). But then there has to be extra assert statements to check that dim 0 matches for all arguments. I can do that, it depends what is better use case.